core/contracts/Governor.sol

```diff
@@ -0,0 +1,139 @@
+     * @param data the transaction data to attach to the call.
+     * @dev You can create the data portion of the transaction by doing the following:
+     * ```
+     * const truffleContractInstance = await TruffleContract.deployed()
+     * const data = truffleContractInstance.methods.methodToCall(arg1, arg2).encodeABI()
+     * ```
+     */
+    function propose(address to, uint value, bytes calldata data) external onlyRoleHolder(uint(Roles.Proposer)) {
+        uint id = proposals.length;
+        uint time = getCurrentTime();
+
+        proposals.push(Proposal({
+            to: to,
+            value: value,
+            data: data,
+            requestTime: time
+        }));
+
+        bytes32 identifier = _constructIdentifier(id);
+
+        // Request a vote on this proposal in the DVM.
+        Voting voting = _getVoting();
+        voting.addSupportedIdentifier(identifier);
+        voting.requestPrice(identifier, time);
+        voting.removeSupportedIdentifier(identifier);
+
+        emit NewProposal(id, to, value, data);
+    }
+
+    /**
+     * @notice Executes a proposed governance action that has been approved by voters. This can be called by anyone.
+     */
+    function executeProposal(uint id) external {
+        Proposal storage proposal = proposals[id];
+        int price = _getVoting().getPrice(_constructIdentifier(id), proposal.requestTime);
+
+        require(price != 0, "Cannot execute, proposal was voted down.");
+        require(_executeCall(proposal.to, proposal.value, proposal.data), "Proposal execution failed.");
+
+        emit ProposalExecuted(id);
+    }
+
+    function numProposals() external view returns (uint) {
+        return proposals.length;
+    }
+
+    function _constructIdentifier(uint id) private pure returns (bytes32 identifier) {
+        bytes32 bytesId = _uintToBytes(id);
+        return _addPrefix(bytesId, "Admin ", 6);
+    }
+
+    function _executeCall(address to, uint256 value, bytes memory data)
+        private
+        returns (bool success)
+    {
+        // Mostly copied from:
+        // solhint-disable-next-line max-line-length
+        // https://github.com/gnosis/safe-contracts/blob/59cfdaebcd8b87a0a32f87b50fead092c10d3a05/contracts/base/Executor.sol#L23-L31
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            let inputData := add(data, 0x20)
+            let inputDataSize := mload(data)
+            success := call(gas, to, value, inputData, inputDataSize, 0, 0)
+        }
+    }
+
+    function _getVoting() private view returns (Voting voting) {
+        return Voting(finder.getImplementationAddress("Oracle"));
+    }
+
+    function _uintToBytes(uint v) private pure returns (bytes32 ret) {
+        if (v == 0) {
+            ret = "0";
+        } else {
+            while (v > 0) {
+                ret = bytes32(uint(ret) / (2 ** 8));
+                ret |= bytes32(((v % 10) + 48) * 2 ** (8 * 31));
```

Comment:
Just curious, is there a reason why you prefer using `/ (2 ** 8)` rather than ` >> 1`?

Reply:
Because I copied this code and didn't wanna mess with it too much. The endianness of bytes32 is different from that of uints, so you have to be a little careful with doing this on different types. I'll add a comment with a link to the original code.

Reply:
Happy to change if you think that would make it more readable though.

Reply:
Added comment with link.

Reply:
LGTM, thanks!

Reply:
After thinking about it a little more, the typecasting makes this more confusing than it needs to be. I modified the logic to use bitshifts and it's definitely more readable.